Restrict classes to the gem namespace

saturnflyer opened the pull request:
## Why

`SOF::Parser` and `SOF::TimeSpan` sat directly under `SOF`, a namespace shared with other libraries and applications. Both carry the comment *"This class is not intended to be referenced directly. This is an internal implementation of Cycle behavior."* An internal implementation detail of `Cycle` has no claim on the shared namespace.

## Summary

- `SOF::Parser` → `SOF::Cycle::Parser`
- `SOF::TimeSpan` → `SOF::Cycle::TimeSpan`

`SOF::Cycles` stays where it is. The gem owns two names under `SOF`, which is a deliberate trade: the handlers each inherit from `SOF::Cycle`, so nesting them beneath it would put a class inside its own superclass. A second top-level constant is the cheaper price.

No aliases are left at the old names. An alias would keep holding the namespace this change exists to release.

## Notes for reviewers

**These are file moves, but git renders the `lib/` ones as add + delete.** The added nesting re-indents every line, so rename detection fails even at 40% similarity. Reviewing with whitespace ignored shows the bodies are untouched.

Low breakage risk: both classes are documented internals. Predicate-based and public-API code (`SOF::Cycle.for`, `SOF::Cycle.legend`, `cycle.end_of?`, `SOF::Cycle::InvalidInput`) is unaffected.

## Review feedback

- **Freeze `dormant_capable_kinds`** — done, in its own commit.
- **Make it a constant instead** — not done, and worth recording why. Handlers register themselves through `Cycle.inherited` as each file in `sof/cycles` loads, which happens *after* `parser.rb`. At that point `Cycle.cycle_handlers` is empty, so a constant would capture `[]` permanently and `dormant_capable?` would return false for every `E`, `I` and `W` cycle. That reasoning is now a comment beside the method so it doesn't get "simplified" later.
- **The `@time_span` shape-variations warning** — done, and extended. The warning is live on Ruby 4.0; enabling it (`-W:performance`) and building a control class that exceeds the limit reproduces it, and Ruby's own remediation text is *"define instance variables in a consistent order, for instance by eagerly defining them all in the #initialize method."* Neither class here was near the eight-variation threshold, so no warning was firing — but `DatePeriod` was the more interesting case than `Parser`: `end_date` and `begin_date` each added an ivar on first call, so the shape depended on call order. Both now declare every ivar in the constructor and take exactly one shape, and the `||= {}` guards are gone.
- **Make `Parser` a module to avoid the nested-class antipattern** — not applicable; `Parser` is instantiated. The related concern about nesting is why `SOF::Cycles` is left alone rather than folded under `Cycle`.

Deferred as out of scope for a namespace change: decoupling from ActiveSupport, dependency injection in `.load` (which must stay a plain dump/load pair for ActiveRecord serialization), casting in the constructor, and extracting a serializer object.

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+  # This class is not intended to be referenced directly.
+  # This is an internal implementation of Cycle behavior.
+  class Cycle
+    class Parser
```
:art: Would it be possible to make this a module instead of a class in order to avoid the [Nested Classes](https://alchemists.io/articles/ruby_antipatterns#_nested_classes) antipattern?

saturnflyer replied:
That would be a good thing, yes. In future changes perhaps

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/hash/keys"
```
:thought_balloon: This is a heavy set of requirements on Rails/ActiveRecord. Would it be possible to decouple this gem from Rails so it's pure Ruby since Rails is so heavyweight for a gem? For example, you the [Refinements](https://alchemists.io/projects/refinements) gem is perfect for situations like this.

saturnflyer replied:
This was extracted from a rails app. Yes, it would be good to limit dependencies but it's not a high priority at the moment

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+        (?<from>F(?<from_date>\d{4}-\d{2}-\d{2}))?$ # optional from
+      /ix
+
+      def self.dormant_capable_kinds = %w[E W]
```
:art: Would a constant be better than a class method since this is what constants are meant for?

:bulb: We should freeze this since there is no additional behavior?

saturnflyer replied:
We should freeze it, yes. 
In my opinion, constants are great for internal reference but methods are better for external reference. 
I'm looking at this on mobile so I don't actually know where this is used but regardless the scope of this PR is namespace changes so that would need to be addressed in the future if necessary

bkuhlmann replied:
> [C]onstants are great for internal reference but methods are better for external reference.

Hmm, that's interesting. I've not heard that before. I suppose I could see that making sense if the class method was made private. At least that would strengthen the *intent* of class method being a *internal* reference.

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+        new(notation_or_parser)
+      end
+
+      def self.load(hash)
```
:art: Would it make sense to inject dependencies for more flexibility? For example, here's a slight refactoring:

``` ruby
def self.load(keys: %i[volume kind period_count period_key], defaults: {volume: 1}, **attributes)
  updated_attributes = defaults.merge! attributes.symbolize_keys!
  text = "V#{updated_attributes.values_at(*keys).join}"

  return new text unless updated_attributes[:from_date]

  new [text, "F#{updated_attributes[:from_date]}"].join
end
```

...but even with this refactoring, there are a few issues:

. The `keys` should be a constant.
. The `defaults` should be a constant.
. The `.load` method might not belong on this class? It feels a little out of place, maybe?

saturnflyer replied:
No, I don't think so. This method is designed to be used in ActiveRecord serialization which depends on a simple dump and load

bkuhlmann replied:
Oh, so for serialization then. That would be a good candidate for serialization objects which would clarify the separation of concerns. Can't tackle that now but would definitely clarify the functionality.

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+      end
+
+      def initialize(notation)
+        @notation = notation&.upcase
```
:art: Could we make our code more confident (channeling Avdi Grimm's _Exceptional Ruby_ book) by casting? Example:

``` ruby
notation = String notation
@notation = notation.upcase
@match = notation.match PARTS_REGEX
```

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+      delegate [:period, :humanized_period] => :time_span
+
+      # Return a TimeSpan object for the period and period_count
+      def time_span
```
:art: I'm pretty sure this will cause a Ruby _shape variations_ performance warning. To avoid this, we can use `@time_span = nil` in the constructor.

:bulb: Here's what I use in my `.bashrc` to stay on top of performance warnings in case it helps: `export RUBYOPT="-W:deprecated -W:performance --yjit --debug-frozen-string-literal"`. You only need `-W:performance` to see the performance warnings but you might want the rest too.

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+      def activated_notation(date)
+        return notation unless dormant_capable?
+
+        self.class.load(to_h.merge(from_date: date.to_date)).notation
```
:thought_balloon: This is interesting. Should we use `#with` instead of `.load`? Then you'd be able to avoid `self.class`.

saturnflyer replied:
Given that the load method is for serialization, no. This is just reusing it.

bkuhlmann commented on lib/sof/cycle/parser.rb:
```diff
@@ -0,0 +1,108 @@
+
+      def ==(other) = other.to_h == to_h
+
+      def to_h
```
:art: If the keys were injected (mentioned earlier in the `.load` method) then you could zip the keys with the values. ...but I think there's something else going on here because it feels like we have an object wanting be born? By this, I mean, we  probably should have a `Data` object which is injected and answered back. This way this responsibility doesn't fall on this class alone?

bkuhlmann replied:
Based on the above conversation I now know what the object is that's hidden in this implementation: it's a serializer.

bkuhlmann reviewed (commented):
Left some feedback on the first class only but I think some of what I'm talking about could be applied to the `TimeSpan` class too.

saturnflyer committed:
Restrict classes to the gem namespace

Parser and TimeSpan sat directly under SOF, so this gem owned three
constants in a namespace other SOF libraries and applications share.
Both are documented as internal implementations of Cycle behavior, so
neither has a claim on it.

They now nest under Cycle, where version.rb already lives. Nothing is
left behind at the old names: an alias would hold the namespace this
change exists to release.

Changed: `SOF::Parser` and `SOF::TimeSpan` are now `SOF::Cycle::Parser` and `SOF::Cycle::TimeSpan`
Version: minor

saturnflyer committed:
Freeze the dormant-capable kinds

The returned array is derived state with no behavior of its own, so
nothing should be mutating it.

It stays a method rather than becoming a constant, and the reason is now
recorded next to it: handlers register through Cycle.inherited as each
file in sof/cycles loads, which happens after this file. A constant
would capture the empty set, and every dormant-capable kind would go
unrecognized.

saturnflyer committed:
Define instance variables eagerly

Parser and DatePeriod each grew instance variables on first use, so their
instances took several object shapes instead of one. DatePeriod was the
worse of the two: end_date and begin_date each added an ivar when called,
so the shape depended on which ran first.

Both now declare every ivar in the constructor, which is what Ruby's own
performance warning recommends. Neither class was close to the eight
variation threshold that triggers it, so this is not a fix for an
observed warning — but the memoization guards were doing shape work for
no reason, and dropping them makes both readers shorter.

Verified with -W:performance: each class now takes exactly one shape.
